Add a federated identity guide

wbamberg opened the pull request:
This PR adds a guide to federated identity.

- an overview of what it is
- a moderately close look at the main flow in OpenID Connect
- a short but sufficient (I think) discussion of the use of third-party cookies in federated identity systems
- an overview of FedCM
- a short section on considerations around choosing IdPs
- summary of strengths and weaknesses

**Note that preview images are broken because of https://github.com/mdn/fred/issues/904.**

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+- The RP passes the code verifier in the _code_verifier_ parameter.
+- The IdP hashes the code verifier, and compares the result with the stored code challenge: if they do not match, then the token request is denied.
+
+This defends against two attacks: CSRF against the RP's redirect URL, and authorization code injection.
```
Link to attack page for CSRF

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+This defends against two attacks: CSRF against the RP's redirect URL, and authorization code injection.
+
+The CSRF attack works as follows:
```
Consider also having the attacks as sub headings. There's a lot going on here and breaking it up might make it more digestible.

wbamberg replied:
I didn't do this because are already at H4(!) but you're right, done in https://github.com/mdn/content/pull/41779/commits/412b8efd124a78fc10cec87c9a9ea19cdf33822f.

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+5. The RP signs the user into the attacker's account: now any information or instructions they provide are under the attacker's control.
+
+Essentially, the attack succeeds because the RP doesn't know that the authentication response is not a response to a request made on behalf of the user. PKCE prevents this attack because the IdP would have a stored code challenge for the attacker's access code, and this would not match the code verifier that the RP passes into the token request.
```
Consider rewording, because the attack doesn't succeed.

hamishwillee replied:
Let me put it another way. After reading this I am not sure how PKCE prevented the attack - I'm just too tired. I might be able to understand if the place at which the request was prevented was in the preceding bullet points. Even if just a note or highlight at the point where success is thwarted.

Same problem with the next case.

I might be wrong. I'll try look at it with fresh eyes next time, not at the end of the day.

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+1. The attacker makes an authentication request to the IdP for themselves, and gets back an authorization code for their own tokens.
+
+2. The attacker tricks the user's browser into making an HTTP request to the RP's redirect URL, including the attacker's authorization code. To the RP, this looks like a response from the IdP to an authentication request originating from the user.
```
Might be worth numbering the point of injection on the preceding flow so it is easier to see what is happening (just thinking aloud)

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+5. The RP signs the attacker into the user's account.
+
+Note that this is really the inverse of the CSRF attack. Note also that the `state` parameter doesn't help here, because the authentication request and response really do belong to the same flow - the attacker's.
```
"Note that this is really the inverse of the CSRF attack." I don't understand in what way.

wbamberg replied:
Removed, I don't think this helps (although I do think it is interesting).

hamishwillee replied:
This caught my eye (i.e. this is just flyby response). I thought this was interesting too. You could add it back, but do it more fully as you did in your extra info:

> (It's the inverse of CSRF, because with CSRF it's the user signing in, but splicing in the attacker's auth code.)

wbamberg commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+PKCE protects against this attack, because the IdP's token endpoint will check the `code_verifier` in the token request against the code challenge that it received in the authentication request.
+
+- Because the IdP's stored code challenge is associated with the authorization code, it will be the challenge for the user.
+- Because the `code_verifier` in the token request is part of the attacker's flow, it will be the verifier for the attacker.
```
> "did Hamish need me to go through that because it was unclear or because he is an idiot"

It's hard. It took me a while. One challenge is that using more words and taking it step by step helps, but I don't want to spend too many words on this. Another is that once you understand it, it makes sense, and it's hard to know if _your explanation_ makes sense to someone else who doesn't understand it yet! I do think I have probably tried to compress it too much here.

wbamberg commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+1. The RP makes a {{httpmethod("POST")}} request to the token endpoint. This request includes the following parameters:
+   - `client_id`: Identifies this RP to the IdP.
+   - `client_secret`: The secret used to authenticate the RP to the token endpoint. The RP could use some alternative mechanism for client authentication, such as TLS client authentication.
```
Yeah, just an opaque value. And yes, everything over HTTPS. I didn't say that because it feels like the default. Worth spelling out do you think?

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -116,51 +121,77 @@ In the authentication request:
 - The RP generates a value that is hard to guess and is specific to this authentication request. This value is called the _code verifier_.
 - The RP creates a {{glossary("hash", "cryptographic hash")}} of the code verifier, and uses it as the `code_challenge` parameter in the authentication request.
-- The IdP stores the code challenge, and associates it with the access code that it returns to the RP.
+- The IdP stores the code challenge, and associates it with the authorization code that it returns to the RP.
```
Mabye "user's authorization code" or similar. Just so that later we already have in our minds that this is associated with the user

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,213 @@
+
+This flow is defined in the [OpenID Connect Core](https://openid.net/specs/openid-connect-core-1_0.html) specification.
+
+As a prerequisite, the RP needs to be known to the IdP: the IdP needs to have an ID for the RP, which is called a client ID, and the RP must be able to authenticate itself to the IdP, for example using a shared secret called a client secret or using some other mechanism such as TLS client authentication.
```
It does make it more readable. I accept your argument on RP, IdP. It's taken me lots of work to get my head around the terms. Really, the problem is that Relying Party is not a really intuitive name, even though it is the standard.

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -116,51 +121,77 @@ In the authentication request:
+7. The RP signs the user into the attacker's account: now any information or instructions they provide are under the attacker's control.
 
-5. The RP signs the user into the attacker's account: now any information or instructions they provide are under the attacker's control.
+Essentially, the attack succeeds because the RP doesn't know that the request to the redirect URL is not a response to a request made on behalf of the user.
```
The whole section is clear/much better. 

FWIW this line still grates because it says the attack succeeds, and isn't it that PKCE is part of the standard and you will never actually see this - i.e. its an explanatory device?

If so, maybe "this attack would succeed"? I'm being pedantic.

wbamberg replied:
Not sure. The flow here (steps 1-7 as described) are what would happen without PKCE. If you are assuming PKCE, then you don't get as far as 6, right? It stops at 5. So without PKCE, "the attack succeeds". I think if you do it all conditionally, it gets complicated, because you have to make 6 and 7 conditional too.

Hm. I pushed https://github.com/mdn/content/pull/41779/commits/7f542f3caf8f19c483e197b1750be21bbe8bdce6 to clarify that these attacks are without PKCE.

> isn't it that PKCE is part of the standard and you will never actually see this

Actually no. PKCE is not part of the standard yet although it will I think be in [OAuth 2.1](https://oauth.net/2.1/). But even then there will be clients around that are not using it, which is partly why (a) it's so complicated and (b) it's so important to have best practice guidelines.

wbamberg commented:
Added a section on FedCM -> https://github.com/mdn/content/pull/41779/commits/afe404f00d5daded08fc54e3b1cbd0e13542242c

Added a bit on PKCE downgrade -> https://github.com/mdn/content/pull/41779/commits/ac6e5ec467823db3659815b8d1061e2616d63426

wbamberg commented:
Recommend public-key crypto for client authentication -> https://github.com/mdn/content/pull/41779/commits/ad5a597cc04f55efcba3000f02ca343b002962bb

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+##### Ensuring that PKCE is used
+
+Although PKCE provides effective protection against the attacks described here, it's an optional part of the protocol. This means that an RP must ensure that its chosen IdP not only supports PKCE, but that it mandates the use of PKCE, refusing the token request if a valid code verifier is not included.
```
I very much like this section.

1. Are there any major IdP that unexpectedly do/do not mandate the use of PKCE? It might be good to hint if there are particular concerns or "no concerns for the major players'.
2. Very nittish, but maybe even more direct - the fact that it is optional is interesting, but the most important bit here is that they need to check that it is mandatory - worth a highlight.

```suggestion
To ensure that PKCE is used, the RP must confirm that its chosen IdP not only supports PKCE, but that it also _mandates_ the use of PKCE — refusing the token request if a valid code verifier is not included.
```

wbamberg replied:
> Are there any major IdP that unexpectedly do/do not mandate the use of PKCE? It might be good to hint if there are particular concerns or "no concerns for the major players'.

I'm not familiar enough with OIDC providers to know this unfortunately - and I expect that this is the kind of information that would go out of date quickly.

One very interesting and relevant difference between PKCE and state/nonce is that with PKCE it's the IdP that does the checking, so as an RP you must trust them to do it, and do it properly. Whereas with state/nonce it's the RP that does the checking, so you are more in control of your own destiny.

In some ways that makes PKCE better, because as a rule IdPs are probably going to be more security-minded than random websites. On the other hand it opens this up downgrade attack.

I considered adding this, but I've already spent a _lot_ of words on PKCE.

hamishwillee replied:
That's cool. Leaving this unresolved in case you want to consider my suggestion in 2.

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+In the token request:
+
+1. The RP makes a {{httpmethod("POST")}} request to the token endpoint. This request includes the following parameters:
```
```suggestion
1. The RP makes a {{httpmethod("POST")}} request to the IdP's token endpoint. This request includes the following parameters:
```

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+   - `response_type`: Always `"code"` if we are using the two-part flow described here, which is the recommended option.
+   - `redirect_uri`: The URL in the RP, that the IdP should redirect to once it has attempted to authenticate the user. This is the URL to which the IdP will deliver the authorization code.
+   - `code_challenge`: A cryptographic {{glossary("hash")}} of a secret specific to this authorization request, which will be used by the token endpoint to ensure that the token request is really the counterpart of this authorization request.
+   - `scope`: A list of strings that specify which sets of user data the RP wishes to access.
```
"Practically" what might be in the scope is not clear. I assume this perhaps means email addresses and such. Not sure whether it matters.

wbamberg replied:
I think I felt like exactly what can go in here is too much detail for this guide, and this is the point where you can go and check the spec (or do what your IdP asked you to do).

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+- The RP passes the code verifier in the _code_verifier_ parameter.
+- The IdP hashes the code verifier, and compares the result with the stored code challenge: if they do not match, then the token request is denied.
+
+This defends against two attacks: [CSRF against the RP's redirect URL](#csrf_against_the_redirect_url), and [authorization code injection](#authorization_code_injection).
```
"This" in "This defends against", is associated with the text above not the feature.

```suggestion
PKCE defends against two attacks: [CSRF against the RP's redirect URL](#csrf_against_the_redirect_url), and [authorization code injection](#authorization_code_injection).
```

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+8. The RP signs the attacker into the user's account.
+
+Note that the `state` parameter doesn't help here, because the authentication request and response really do belong to the same flow - the attacker's.
```
Maybe "using the OIDC `state`..." or 
```suggestion
Note that using a `state` parameter doesn't help in this case, because the authentication request and response really do belong to the same flow - the attacker's.
```

wbamberg commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+The [Federated Credential Management API (FedCM API)](/en-US/docs/Web/API/FedCM_API) provides built-in browser support for federated identity. The API does not yet have cross-browser support and is still being actively developed, so we can't fully recommend its use, but it promises several benefits over implementing a protocol like OpenID Connect directly:
+
+- In the OIDC flow we've previously described, the website using OIDC (that is, the RP) has to coordinate the interactions between itself, the user, and the IdP. As we've seen, this is complicated and error-prone. With FedCM, the browser takes care of this interaction: as an RP, you call a browser API, and the browser locates the IdP, asks the user to authenticate, and returns a token from the Id ...
```
Yeah it is confusing. These are unfortunately the terms used in the spec (I feel like when the spec was written it was assumed that this might be an actual ID token). We could use something other than "token" here though - 'object", "value". I don't think we have to use the spec language.

wbamberg replied:
FWIW the spec says:

> The content of the [token](https://w3c-fedid.github.io/FedCM/#dom-identityassertionresponse-token) is opaque to the user agent and can contain anything that the [IDP](https://w3c-fedid.github.io/FedCM/#idp) would like to pass to the [RP](https://w3c-fedid.github.io/FedCM/#rp) to facilitate the login.

see also my related complaints at https://github.com/mdn/content/issues/42025.

hamishwillee replied:
Yuck. I'd be tempted to keep token but make it clear that it is not the same token - because the terminology is both accurate and also is probably used in the API description and would be hard to purge. So "something like"

```suggestion
- In the OIDC flow we've previously described, the website using OIDC (that is, the RP) has to coordinate the interactions between itself, the user, and the IdP. As we've seen, this is complicated and error-prone. With FedCM, the browser takes care of this interaction: as an RP, you call a browser API, and the browser locates the IdP, asks the user to authenticate, and returns an opaque token from the IdP that the RP can use to sign the user in. In an OIDC implementation this token might be the authentication code from the IdP.
```

wbamberg commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -116,51 +121,77 @@ In the authentication request:
-In the authorization code injection attack:
+##### Authorization code injection
+
+In an authorization code injection attack, the attacker steals an authorization code from the target user, and is able to inject it into the attacker's own sign-in flow. The result is that the attacker is signed into the user's account.
```
I think I've addressed this with my "If PKCE were not used," edits.

hamishwillee commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+The specification considers that clients running on the user's browser are public clients, for the same reason we've already encountered: it's too easy for an attacker to access secrets in a browser via attacks such as XSS. Clients running on a web _server_ are confidential clients.
+
+In OIDC, only confidential clients may use client authentication, because only confidential clients can be trusted to maintain the security of the client's credentials.
```
A matter of taste :-)

wbamberg commented on files/en-us/web/security/authentication/federated_identity/index.md:
```diff
@@ -0,0 +1,302 @@
+
+- The next most secure pattern is one in which the website uses a web server to handle all OAuth/OIDC interactions, but then returns the access token to the front end, and the front end then makes API requests directly. In this scenario the website can be a confidential client but malicious code running in the browser (for example, through an XSS attack) can potentially steal access tokens. Howev ...
+
+- The least secure pattern is one in which OAuth/OIDC interactions and interactions with APIs both take place in the front end. This, for example, would be the natural architecture for a {{glossary("SPA", "Single-page app")}}, where the entire application executes in the browser. In this architecture the RP can't be a confidential client, because it can't reliably keep a client secret. This means ...
```
-> https://github.com/mdn/content/pull/41779/commits/8e7a48f3a3003a490f7fbb734e3e248fd860d4de

wbamberg commented:
This has been open for a while and has had some review, so I'm merging it. Happy to handle any more comments in follow-ups. Thanks for your work on the review @hamishwillee , much appreciated.
